src/nupic/algorithms/Anomaly.cpp

```diff
@@ -42,92 +35,18 @@ namespace anomaly {
 Real computeRawAnomalyScore(const SDR& active,
                             const SDR& predicted) {
 
+  NTA_ASSERT(active.dimensions == predicted.dimensions); 
+
   // Return 0 if no active columns are present
   if (active.getSum() == 0) {
     return 0.0f;
   }
 
-  NTA_CHECK(active.dimensions == predicted.dimensions);
-
   // Calculate and return percent of active columns that were not predicted.
   SDR both(active.dimensions);
   both.intersection(active, predicted);
 
   return (active.getSum() - both.getSum()) / Real(active.getSum());
 }
 
-Real computeRawAnomalyScore(vector<UInt>& active,
-                            vector<UInt>& predicted)
```

Comment:
Can we keep this overload?

Reply:
> This is where I keep my notes about how the SDR class should be doing set intersections.

We could keep it, 
or keep the code just as a commet, if you're not actively developing on that? 
And ideally move the comment to SDR.intersection() ?

Reply:
Ok, I we can drop this method, but the rest of my criticism stands.
- Friend classes are confusing and usually not necessary.
- This doesn't include anomaly likelihood.
- External predictive inputs don't work?  I read through the code and it looks like they should just work, if you remove the checks that disable it.

Reply:
> Instead of removing the anomaly class and adding a different class, just reuse the existing anomaly class. 

I thought you wanted me to get rid of the old Anomaly class and replace it with as simple as possibleimplementation in TM (?), which I ended up liking. 

1. I'm OK to revive the old Anomaly class, with a few changes *). 
2. Or I can just move the code TMAnomaly to Anomaly.hpp and be done with it. 
Which of those would you prefer? 

> Friend classes are confusing and usually not necessary. 

I'll avoid 'friend access' when moving to a separate new/old file. 

> This doesn't include anomaly likelihood. 

Yes, and it is intended. And about other refactoring in Anomaly class *): 
- I've implemented Anomaly class in the past to simplify various options with anomaly (thresholds, likelihood, moving average,...) 
- now with later usage experience, I don;t think it's a good design, and as those features can be implemented as "one liners" with current code, I don't think it should be provided with anomaly  class. "Let TM.getAnomalyScore() be the simple raw anomaly" (#A), just conveniently computed. And let the user do what they need later with it. 
  - `thereshold`: simple `if(score > XXX) return 1 else return 0;`
  - `likelihood`: 
    - (probably even does not truly work) 
    - in Anomaly had a FIXME note that it's not truly configurable. And should be separated. 
    - I could use the approach with abstract class `AbstractAnomaly::compute(SDR, SDR)`, but it's complicating things, see argument #A
    - AnomalyLikelihood is not even a true anomaly (computed from 2 SDRs), but a likelihood of anomalies, thus computed AFTER and from anomaly score. So `score = TM.getAnomalyScore(); Likelihood::anomalyProbability(score, timestep);` 
  - `moving average`: again, simple `MovingAverage ma(5); ma.add(score); ma.get();`

TL;DR: the proper configuration and ordering is too complicated in a wrapper Anomaly class, while relatively easy hand-tailored by user (with tools we provide). 

> External predictive inputs don't work? I read through the code and it looks like they should just work, if you remove the checks that disable it.

Externals break it in the way the they add "cell indices out of scope of the current TM" (idx in <numberOfCells() , numOfCells + extra>). 
See for yourself in SDR::setSparseInplace(), you must use Debug build to trigger, that's why it's failing only on the OSX CI. I'm proposing to proceed as is, and discuss the fixes on external inputs in a separate issue or #442

Reply:
> What would you prefer?

My goal is to have convenient way to print out everything interesting about the TM to a human readable string, and from it to determine if the TM is working.  I see the anomaly metric as a diagnostic tool.
 * I want to know the mean & standard deviation of the anomaly.  The anomaly likelihood also uses this info.
 * It should be enabled by default. 

> The proper configuration and ordering is too complicated.

Agreed,
 * The user should be responsible for discarding anomaly readings when they haven't finished training the network.
 * The user should be responsible for applying the threshold.
* AnomalyLikelihood needs 1 parameter to do what it does, and it has 4 or 5.